src/protocol/securemul.rs

```diff
@@ -14,6 +16,53 @@ pub struct DValue<F> {
     d: F,
 }
 
+#[derive(Debug)]
+pub struct SecureMulAll<'a, N> {
+    prss: &'a PrssSpace,
+    gateway: &'a Gateway<N>,
+    step: &'a UniqueStepId,
+}
+
+impl<'a, N: Network> SecureMulAll<'a, N> {
+    pub fn new(prss: &'a PrssSpace, gateway: &'a Gateway<N>, step: &'a UniqueStepId) -> Self {
+        Self {
+            prss,
+            gateway,
+            step,
+        }
+    }
+
+    pub async fn execute<F: Field>(
+        self,
+        a: Batch<Replicated<F>>,
+        b: Batch<Replicated<F>>,
```

Comment:
Should I take slices instead of clones to avoid extra memory allocations?

Reply:
this code moves `a` and `b`, so there is no extra allocations here. On the caller's side there might be a need to clone `a` and `b` though. I like your suggestion of taking a reference here because it is more appropriate here. 

There might be an implementation that takes `a` by value (or exclusive reference) and borrows `b` to then re-use `a` to store the result of `a` * `b`. This makes the whole operation  $2 \times N$ in space compared to $3 \times N$

Reply:
Shouldn't this take `impl IntoIterator<Item = Replicated<F>>` or maybe just a slice?  For the code that you are working on this will always be a finite set of inputs, likely a slice.

Reply:
I changed the parameter to take a slice, to allow the call site to continue using the referenced vector.